numba/targets/base.py

```diff
@@ -34,25 +35,56 @@ class OverloadSelector(object):
 
     In the current implementation:
     - a "signature" is a tuple of type classes or type instances
-    - the "best candidate" is simply the first match
+    - the "best candidate" is the most specific match
     """
 
     def __init__(self):
         # A list of (formal args tuple, value)
         self.versions = []
 
     def find(self, sig):
-        for ver_sig, impl in self.versions:
-            if ver_sig == sig:
-                return impl
+        candidates = dict(self._select_compatible(sig))
+        if candidates:
+            ranked = self._sort_signatures(candidates.keys())
+            return candidates[ranked[0]]
+        else:
+            raise NotImplementedError(self, sig)
 
-            # As generic type
+    def _select_compatible(self, sig):
+        """
+        Select all compatible signatures and their implementation.
+        """
+        out = {}
+        for ver_sig, impl in self.versions:
             if self._match_arglist(ver_sig, sig):
-                return impl
-
-        raise NotImplementedError(self, sig)
+                out[ver_sig] = impl
+        return out
+
+    def _sort_signatures(self, candidates):
+        """
+        Sort signatures in ascending level of genericity.
+        """
+        scoring = defaultdict(int)
+        for this, other in permutations(candidates, r=2):
+            matched = self._match_arglist(formal_args=this, actual_args=other)
```

Comment:
I think using `_match_arglist` is wrong, since it expects `actual_args` to hold type *instances*, but here you're passing it a formal signature, there most probably type classes.